test/dynamo/test_higher_order_ops.py

```diff
@@ -654,6 +654,229 @@ def fn(x):
+
+            def fn(x):
+                return (x.sin() + y).sum()
+
+            return torch.func.grad(fn)(x)
+
+        x = torch.randn(3, 3, 3)
+        self._grad_compile_check(
+            wrapper_fn,
+            x,
+            expected_module_names=("grad_body_0",),
+            expected_node_names=("contiguous",),
+        )
+
+    def test_grad_has_aux(self):
+        counters.clear()
+
+        def wrapper_fn(x):
+            y = 3.14
+
+            def fn(x):
+                return ((x.sin() + y).sum(), x.cos())
+
+            return torch.func.grad(fn, has_aux=True)(x)
+
+        x = torch.randn(3, 3, 3)
+        self._grad_compile_check(
+            wrapper_fn,
+            x,
+            expected_module_names=("grad_body_0",),
+            expected_node_names=("contiguous",),
+        )
+
+    def test_grad_two_tensor_has_aux(self):
+        counters.clear()
+
+        def fn(x, y):
+            return ((x.sin() + y).sum(), x.cos())
+
+        def wrapper_fn(x, y):
+            return torch.func.grad(fn, has_aux=True)(x, y)
+
+        y = torch.randn(3, 3, 3)
+        x = torch.randn(3, 3, 3)
+        self._grad_compile_check(
+            wrapper_fn,
+            x,
+            y,
+            expected_module_names=("grad_body_0",),
+            expected_node_names=("contiguous",),
+        )
+
+    def test_grad_two_tensor_all_grad_has_aux(self):
+        counters.clear()
+
+        def fn(x, y):
+            return ((x.sin() + y).sum(), x.cos())
+
+        def wrapper_fn(x, y):
+            return torch.func.grad(fn, argnums=(0, 1), has_aux=True)(x, y)
+
+        y = torch.randn(3, 3, 3)
+        x = torch.randn(3, 3, 3)
+        self._grad_compile_check(
+            wrapper_fn,
+            x,
+            y,
+            expected_module_names=("grad_body_0",),
+            expected_node_names=("contiguous", "contiguous_1"),
+        )
+
+    def test_grad_over_grad(self):
+        counters.clear()
+
+        def fn(x):
+            return x.sin().sum()
+
+        def wrapper_fn(x):
+            return torch.func.grad(torch.func.grad(fn))(x)
+
+        x = torch.randn(())
+        self._grad_compile_check(
+            wrapper_fn,
+            x,
+            expected_module_names=("grad_body_1.grad_body_0", "grad_body_1"),
+        )
+
+    def test_grad_with_graph_break(self):
+        counters.clear()
+
+        def fn(x):
+            torch._dynamo.graph_break()
+            return x.sin().sum()
+
+        def wrapper_fn(x):
+            return torch.func.grad(fn)(x)
+
+        x = torch.randn(3, 3, 3)
+        actual = wrapper_fn(x)
+        expected = torch.compile(wrapper_fn, backend="aot_eager", fullgraph=False)(x)
+        self.assertEqual(actual, expected)
+
+    def test_grad_with_side_effect(self):
+        counters.clear()
+
+        foo = [1, 2]
+
+        def fn(x):
+            foo.append(3)
+            return x.sin().sum()
+
+        def wrapper_fn(x):
+            return torch.func.grad(fn)(x)
```

Comment:
I don't think this will work out of the box now as 
```
torch._dynamo.exc.Unsupported: HigherOrderOperator with body that accepts non-Tensors as input
```

Will add a test to make sure we graph-break for now.

Thanks!